File: examples/python/dex_examples/products/order-processing/order_processing_flow.py
# Copyright (c) 2022-2026 Super Durable, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from datetime import timedelta

from dex import (
    Attribute,
    AttributeIndex,
    Channel,
    Context,
    Flow,
    IndexType,
    PersistenceSchema,
    RPCResult,
    RetryPolicy,
    Step,
    StepDecision,
    StepList,
    StepOptions,
    Timer,
    Wait,
    go_to,
    graceful_complete,
    rpc,
)

from dex_examples.shared.my_dependency_service import MyDependencyService
from dex_examples.products.order_processing.order_request import OrderRequest

SELLER_REMINDER_TIMER = "seller-reminder"
SHIP_RETRY = RetryPolicy(
    initial_interval=timedelta(seconds=1),
    maximum_attempts=2,
)
CHARGE_RETRY = RetryPolicy(maximum_attempts=3)
REFUND_RETRY = RetryPolicy(maximum_attempts=3)


class ChargeStep(Step[OrderRequest]):
    def __init__(self, flow: "OrderProcessingFlow") -> None:
        self.flow = flow

    def get_step_options(self) -> StepOptions:
        return StepOptions(execute_retry=CHARGE_RETRY)

    def execute(self, context: Context, input: OrderRequest) -> StepDecision:
        self.flow.service.charge_user(input.email, input.customer_id, input.amount)
        self.flow.order_status.set(context, "charged")
        return go_to(self.flow.ship, input)


class ShipStep(Step[OrderRequest]):
    def __init__(self, flow: "OrderProcessingFlow") -> None:
        self.flow = flow

    def get_step_options(self) -> StepOptions:
        return StepOptions(execute_retry=SHIP_RETRY).on_execute_failure_proceed_to(
            self.flow.refund,
            StepOptions(execute_retry=REFUND_RETRY),
        )

    def wait_for(self, context: Context, input: OrderRequest) -> Wait:
        del context, input
        return Wait.any_of(
            self.flow.seller_ok.for_one(),
            Timer.by_duration(timedelta(hours=24), condition_id=SELLER_REMINDER_TIMER),
        )

    def execute(self, context: Context, input: OrderRequest) -> StepDecision:
        if context.has_timer_fired():
            self.flow.service.send_email(
                input.email,
                "Reminder: approve shipment",
                "Please approve or provide a tracking number.",
            )
            return go_to(self, input)
        if input.fail_ship:
            raise RuntimeError(f"ship failed for order {input.order_id}")
        self.flow.service.update_external_system(f"ship {input.order_id}")
        self.flow.order_status.set(context, "shipped")
        return graceful_complete(f"shipped:{input.order_id}")


class RefundStep(Step[OrderRequest]):
    def __init__(self, flow: "OrderProcessingFlow") -> None:
        self.flow = flow

    def execute(self, context: Context, input: OrderRequest) -> StepDecision:
        self.flow.service.update_external_system(f"refund {input.order_id}")
        self.flow.order_status.set(context, "refunded")
        return graceful_complete(f"refunded:{input.order_id}")


class OrderProcessingFlow(Flow[OrderRequest]):
    order_status = Attribute(
        "order-status",
        str,
        AttributeIndex(IndexType.KEYWORD),
    )
    seller_ok = Channel[str]("seller-ok", str)

    def __init__(self, service: MyDependencyService) -> None:
        self.service = service
        self.charge = ChargeStep(self)
        self.ship = ShipStep(self)
        self.refund = RefundStep(self)

    def get_steps(self) -> StepList[OrderRequest]:
        return StepList.start_step(self.charge).other_steps(self.ship, self.refund)

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(self.order_status, self.seller_ok)

    @rpc
    def approve(self, context: Context, _note: str) -> RPCResult[str]:
        self.seller_ok.publish(context, "approved")
        return RPCResult("ok")

    @rpc
    def describe(self, context: Context) -> RPCResult[str]:
        return RPCResult(self.order_status.get(context))

File: examples/java/src/main/java/io/superdurable/dex/products/orderprocessing/OrderProcessingFlow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.superdurable.dex.products.orderprocessing;

import io.superdurable.dex.Attribute;
import io.superdurable.dex.AttributeIndex;
import io.superdurable.dex.Channel;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;
import io.superdurable.dex.RetryPolicy;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepOptions;
import io.superdurable.dex.Timer;
import io.superdurable.dex.Wait;
import io.superdurable.dex.shared.MyDependencyService;
import org.springframework.stereotype.Component;

import java.time.Duration;

@Component
public class OrderProcessingFlow implements Flow<OrderRequest> {
    public final Attribute<String> orderStatus = Attribute.define(
            "order-status",
            String.class,
            new AttributeIndex(AttributeIndex.Type.KEYWORD));
    public final Channel<String> sellerOk = Channel.define("seller-ok", String.class);

    private final MyDependencyService service;
    private final ChargeStep charge;
    private final ShipStep ship;
    private final RefundStep refund;

    public OrderProcessingFlow(final MyDependencyService service) {
        this.service = service;
        this.charge = new ChargeStep(service);
        this.ship = new ShipStep(service);
        this.refund = new RefundStep(service);
    }

    @Override
    public StepList<OrderRequest> getSteps() {
        return StepList.startStep(charge).otherSteps(ship, refund);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(orderStatus, sellerOk);
    }

    @RPC
    public RPCResult<String> approve(final Context context, final String note) {
        sellerOk.publish(context, "approved");
        return RPCResult.of("ok");
    }

    @RPC
    public RPCResult<String> describe(final Context context) {
        return RPCResult.of(orderStatus.get(context));
    }

    final class ChargeStep implements Step<OrderRequest> {
        private final MyDependencyService service;

        ChargeStep(final MyDependencyService service) {
            this.service = service;
        }

        @Override
        public String getStepType() {
            return "ChargeStep";
        }

        @Override
        public Class<OrderRequest> getInputType() {
            return OrderRequest.class;
        }

        @Override
        public StepOptions getStepOptions() {
            return StepOptions.newBuilder()
                    .executeRetry(RetryPolicy.newBuilder()
                            // .totalDuration(Duration.ofHours(1))
                            .totalDuration(Duration.ofSeconds(3))
                            .build())
                    .build();
        }

        @Override
        public StepDecision execute(final Context context, final OrderRequest order) {
            service.chargeUser(order.email, order.customerId, order.amount);
            orderStatus.set(context, "charged");
            return StepDecision.goTo(ship, order);
        }
    }

    final class ShipStep implements Step<OrderRequest> {
        private final MyDependencyService service;

        ShipStep(final MyDependencyService service) {
            this.service = service;
        }

        @Override
        public String getStepType() {
            return "ShipStep";
        }

        @Override
        public Class<OrderRequest> getInputType() {
            return OrderRequest.class;
        }

        @Override
        public StepOptions getStepOptions() {
            return StepOptions.newBuilder()
                    .executeRetry(RetryPolicy.newBuilder()
                            // .totalDuration(Duration.ofHours(1))
                            .totalDuration(Duration.ofSeconds(3))
                            .build())
                    .onExecuteFailureProceedTo(
                            refund,
                            StepOptions.newBuilder()
                                    .executeRetry(RetryPolicy.newBuilder()
                                            // .totalDuration(Duration.ofHours(1))
                                            .totalDuration(Duration.ofSeconds(3))
                                            .build())
                                    .build())
                    .build();
        }

        @Override
        public Wait waitFor(final Context context, final OrderRequest order) {
            return Wait.anyOf(
                    sellerOk.forOne(),
                    Timer.byDuration(Duration.ofHours(24)));
        }

        @Override
        public StepDecision execute(final Context context, final OrderRequest order) {
            if (context.hasTimerFired()) {
                service.sendEmail(
                        order.email,
                        "Reminder: approve shipment",
                        "Please approve or provide a tracking number.");
                return StepDecision.goTo(ship, order);
            }
            service.shipItem(order.orderId, order.testFailAtShipping);
            orderStatus.set(context, "shipped");
            return StepDecision.gracefulComplete("shipped:" + order.orderId);
        }
    }

    final class RefundStep implements Step<OrderRequest> {
        private final MyDependencyService service;

        RefundStep(final MyDependencyService service) {
            this.service = service;
        }

        @Override
        public String getStepType() {
            return "RefundStep";
        }

        @Override
        public Class<OrderRequest> getInputType() {
            return OrderRequest.class;
        }

        @Override
        public StepDecision execute(final Context context, final OrderRequest order) {
            service.updateExternalSystem("refund " + order.orderId);
            orderStatus.set(context, "refunded");
            return StepDecision.gracefulComplete("refunded:" + order.orderId);
        }
    }
}
